src/mediapipe_calculators/python_executor_calculator.cc

```diff
@@ -0,0 +1,160 @@
+
+/* 
+TODO: Streaming support:
+    - timestamping
+    - loopback input
+*/ 
+
+std::shared_ptr<PythonNodeResource> nodeResources;
+std::unique_ptr<PyObjectWrapper<py::iterator>> pyIteratorPtr;
+
+void pushOutputs(CalculatorContext* cc, py::list pyOutputs) {
+    py::gil_scoped_acquire acquire;
+    for (const std::string& tag : cc->Outputs().GetTags()) {
+        for (py::handle pyOutputHandle : pyOutputs) {
+            py::object pyOutput = pyOutputHandle.cast<py::object>();
+            nodeResources->pythonBackend->validateOvmsPyTensor(pyOutput);
+            if (pyOutput.attr("name").cast<std::string>() == tag) {
+                std::unique_ptr<PyObjectWrapper<py::object>> outputPtr = std::make_unique<PyObjectWrapper<py::object>>(pyOutput);
+                cc->Outputs().Tag(tag).Add(outputPtr.release(), cc->InputTimestamp());
+            }
+        }
+    }
+}
+
+public:
+    static absl::Status GetContract(CalculatorContract* cc) {
+        LOG(INFO) << "PythonExecutorCalculator [Node: " << cc->GetNodeName() << "] GetContract start";
+        RET_CHECK(!cc->Inputs().GetTags().empty());
+        RET_CHECK(!cc->Outputs().GetTags().empty());
+        for (auto& input : cc->Inputs()) {
+            input.Set<PyObjectWrapper<py::object>>();
+        }
+        for (auto& output : cc->Outputs()) {
+            output.Set<PyObjectWrapper<py::object>>();
+        }
+        cc->InputSidePackets().Tag(INPUT_SIDE_PACKET_TAG).Set<PythonNodesResources>();
+        LOG(INFO) << "PythonExecutorCalculator [Node: " << cc->GetNodeName() << "] GetContract end";
+        return absl::OkStatus();
+    }
+
+    absl::Status Close(CalculatorContext* cc) final {
+        LOG(INFO) << "PythonExecutorCalculator [Node: " << cc->NodeName() << "] Close";
+        return absl::OkStatus();
+    }
+
+    absl::Status Open(CalculatorContext* cc) final {
+        LOG(INFO) << "PythonExecutorCalculator [Node: " << cc->NodeName() << "] Open start";
+        PythonNodesResources nodesResourcesPtr = cc->InputSidePackets().Tag(INPUT_SIDE_PACKET_TAG).Get<PythonNodesResources>();
+        auto it = nodesResourcesPtr.find(cc->NodeName());
+        if (it == nodesResourcesPtr.end()) {
+            LOG(INFO) << "Could not find initialized Python node named: " << cc->NodeName();
+            RET_CHECK(false);
+        }
+        nodeResources = it->second;
+        LOG(INFO) << "PythonExecutorCalculator [Node: " << cc->NodeName() << "] Open end";
+        return absl::OkStatus();
+    }
+
+    absl::Status Process(CalculatorContext* cc) final {
+        LOG(INFO) << "PythonExecutorCalculator [Node: " << cc->NodeName() << "] Process start";
+        py::gil_scoped_acquire acquire;
+        try {
+            py::print("PYTHON: Acquired GIL");
+            if (pyIteratorPtr) { // Iterator initialized
+                py::print("PyIterator initialized block");
+                if (pyIteratorPtr->getImmutableObject() != py::iterator::sentinel()) {
```

Comment:
Why do we need getImmutable and getMutable? Cant we use mutable in both scenarios?

Reply:
It's for the generator object. We need const for getting the data of checking if we're done generating new data. We need non-const for incrementing the iterator.